src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -353,16 +352,14 @@ void executeQueryWithParallelReplicas(
         // shard_num is 1-based, but getClusterWithSingleShard expects 0-based index
         auto single_shard_cluster = not_optimized_cluster->getClusterWithSingleShard(shard_num - 1);
         // convert cluster to representation expected by parallel replicas
-        new_cluster = single_shard_cluster->getClusterWithReplicasAsShards(settings);
+        new_cluster = single_shard_cluster->getClusterWithReplicasAsShards(settings, settings.max_parallel_replicas);
     }
     else
     {
-        new_cluster = not_optimized_cluster->getClusterWithReplicasAsShards(settings);
+        new_cluster = not_optimized_cluster->getClusterWithReplicasAsShards(settings, settings.max_parallel_replicas);
```

Comment:
If `max_parallel_replicas` less than the number of replicas in cluster, then replicas will be shuffled before converting to shards, [here](https://github.com/ClickHouse/ClickHouse/blob/500a281acd6eb764dd6469e3249c87fd1139ebf7/src/Interpreters/Cluster.cpp#L774). So, every query execution will assign different nodes. In context, of cache locality, it's not what we want

Reply:
for the history: we agreed to leave it this way until we have a better solution. because the other alternative is to always peek the same prefix of nodes and leave the rest not utilised at all.